File: data_pipeline/fusion/kalman.py
"""Multi-timescale Kalman filter bank for price fusion."""
import numpy as np
from typing import Dict, Tuple, Optional
from datetime import datetime

from data_pipeline.config import KALMAN_FILTERS


class KalmanFilter:
    """Single Kalman filter for price estimation."""

    def __init__(self, process_noise: float, observation_noise: float):
        """Initialize 2D Kalman filter (price + velocity).

        Args:
            process_noise: Process noise covariance (Q)
            observation_noise: Observation noise covariance (R)
        """
        # State: [price, velocity]
        self.x = np.array([0.0, 0.0])  # Initial state
        self.P = np.eye(2) * 1000.0  # Initial covariance (high uncertainty)

        # State transition matrix (constant velocity model)
        self.F = np.array([[1.0, 1.0], [0.0, 1.0]])

        # Observation matrix (we observe only price)
        self.H = np.array([[1.0, 0.0]])

        # Process noise covariance
        self.Q = np.eye(2) * process_noise

        # Observation noise covariance
        self.R = np.array([[observation_noise]])

        self.initialized = False

    def predict(self) -> Tuple[float, float]:
        """Prediction step.

        Returns:
            (predicted_price, uncertainty)
        """
        # Predict state
        self.x = self.F @ self.x

        # Predict covariance
        self.P = self.F @ self.P @ self.F.T + self.Q

        return self.x[0], self.P[0, 0]

    def update(self, observation: float, trust: float = 1.0) -> Tuple[float, float, float]:
        """Update step.

        Args:
            observation: Observed price
            trust: Trust score for this observation (0-1)

        Returns:
            (filtered_price, uncertainty, innovation)
        """
        if not self.initialized:
            # Initialize state with first observation
            self.x[0] = observation
            self.initialized = True
            return observation, self.P[0, 0], 0.0

        # Compute innovation
        y = observation - (self.H @ self.x)[0]

        # Innovation covariance (weighted by trust)
        S = (self.H @ self.P @ self.H.T + self.R / trust)[0, 0]

        # Kalman gain
        K = (self.P @ self.H.T) / S

        # Update state
        self.x = self.x + K.flatten() * y

        # Update covariance
        self.P = (np.eye(2) - K @ self.H) @ self.P

        return self.x[0], self.P[0, 0], abs(y)


class KalmanFilterBank:
    """Bank of Kalman filters at different timescales."""

    def __init__(self):
        """Initialize filter bank."""
        self.filters: Dict[str, KalmanFilter] = {}
        for name, params in KALMAN_FILTERS.items():
            self.filters[name] = KalmanFilter(
                process_noise=params["process_noise"],
                observation_noise=params["observation_noise"],
            )

        # Trust scores for each source
        self.trust_scores: Dict[str, float] = {}

    def init_trust(self, source: str, initial_trust: float = 0.5) -> None:
        """Initialize trust score for a source."""
        self.trust_scores[source] = max(0.05, min(0.95, initial_trust))

    def update_trust(self, source: str, innovation: float, uncertainty: float) -> float:
        """Update trust score based on innovation residual.

        Args:
            source: Source name
            innovation: Absolute innovation from Kalman filter
            uncertainty: Current uncertainty estimate

        Returns:
            Updated trust score
        """
        if source not in self.trust_scores:
            self.init_trust(source)

        # Compute z-score of innovation
        if uncertainty > 0:
            z_score = innovation / np.sqrt(uncertainty)
        else:
            z_score = 0.0

        # Update trust
        if z_score < 1.0:
            # Good agreement -> increase trust
            self.trust_scores[source] = min(0.95, self.trust_scores[source] + 0.01)
        elif z_score > 3.0:
            # High disagreement -> decrease trust
            self.trust_scores[source] = max(0.05, self.trust_scores[source] - 0.05)

        return self.trust_scores[source]

    def fuse(
        self,
        observations: Dict[str, float],
        timestamp: datetime,
    ) -> Tuple[float, float, Dict[str, float], bool]:
        """Fuse multiple observations via filter bank.

        Args:
            observations: Dict mapping source -> price
            timestamp: Current timestamp

        Returns:
            (fused_price, confidence, source_weights, anomaly_flag)
        """
        # Initialize trust for new sources
        for source in observations:
            if source not in self.trust_scores:
                self.init_trust(source)

        # Predict step for all filters
        predictions = {}
        for name, filt in self.filters.items():
            pred_price, pred_unc = filt.predict()
            predictions[name] = (pred_price, pred_unc)

        # Update each filter with trusted observations
        filter_outputs = {}
        source_innovations = {}

        for source, obs in observations.items():
            trust = self.trust_scores[source]

            for name, filt in self.filters.items():
                filtered_price, uncertainty, innovation = filt.update(obs, trust)
                if name not in filter_outputs:
                    filter_outputs[name] = []
                filter_outputs[name].append((filtered_price, uncertainty))

                # Track innovation for trust update
                source_innovations[source] = (innovation, uncertainty)

        # Update trust scores
        for source, (innov, unc) in source_innovations.items():
            self.update_trust(source, innov, unc)

        # Fuse filter outputs (weighted by inverse uncertainty)
        fused_prices = []
        weights = []

        for name, outputs in filter_outputs.items():
            # Average across all observations for this filter
            avg_price = np.mean([p for p, u in outputs])
            avg_unc = np.mean([u for p, u in outputs])

            if avg_unc > 0:
                weight = 1.0 / avg_unc
            else:
                weight = 1.0

            fused_prices.append(avg_price * weight)
            weights.append(weight)

        # Final fused price
        total_weight = sum(weights)
        if total_weight > 0:
            fused_price = sum(fused_prices) / total_weight
            confidence = total_weight  # Higher weight = higher confidence
        else:
            fused_price = np.mean(list(observations.values()))
            confidence = 0.1

        # Anomaly detection: check if any source diverges >3σ
        anomaly_flag = False
        for source, obs in observations.items():
            if abs(obs - fused_price) > 3.0 * np.sqrt(1.0 / confidence):
                anomaly_flag = True
                break

        # Source weights (normalized trust scores)
        total_trust = sum(self.trust_scores[s] for s in observations)
        if total_trust > 0:
            source_weights = {s: self.trust_scores[s] / total_trust for s in observations}
        else:
            source_weights = {s: 1.0 / len(observations) for s in observations}

        return fused_price, confidence, source_weights, anomaly_flag

File: dominion_cpp/src/fusion/kalman.cpp
#include "dominion/fusion.hpp"
#include <cmath>
#include <algorithm>
#include <numeric>

namespace dominion {

KalmanFilter::KalmanFilter(double process_noise, double observation_noise)
    : q_(process_noise), r_(observation_noise) {}

void KalmanFilter::predict() {
    if (!state_.initialized) return;

    // State transition: x = F * x
    // F = [[1, 1], [0, 1]]  (constant velocity model)
    double new_price = state_.x[0] + state_.x[1];
    double new_velocity = state_.x[1];
    state_.x[0] = new_price;
    state_.x[1] = new_velocity;

    // Covariance update: P = F * P * F^T + Q
    // Q = [[q, 0], [0, q]]
    auto& P = state_.P;
    double p00 = P[0][0] + 2 * P[0][1] + P[1][1] + q_;
    double p01 = P[0][1] + P[1][1];
    double p10 = p01;
    double p11 = P[1][1] + q_;
    P[0][0] = p00;
    P[0][1] = p01;
    P[1][0] = p10;
    P[1][1] = p11;
}

void KalmanFilter::update(double observation, double trust_weight) {
    if (!state_.initialized) {
        state_.x[0] = observation;
        state_.x[1] = 0.0;
        state_.initialized = true;
        return;
    }

    // Innovation: y = z - H * x
    // H = [[1, 0]]  (observe price only)
    double innovation = observation - state_.x[0];

    // Innovation covariance: S = H * P * H^T + R
    double S = state_.P[0][0] + r_ / trust_weight;

    // Kalman gain: K = P * H^T * S^-1
    double k0 = state_.P[0][0] / S;
    double k1 = state_.P[1][0] / S;

    // State update: x = x + K * y
    state_.x[0] += k0 * innovation;
    state_.x[1] += k1 * innovation;

    // Covariance update: P = (I - K * H) * P
    auto& P = state_.P;
    double p00 = (1.0 - k0) * P[0][0];
    double p01 = (1.0 - k0) * P[0][1];
    double p10 = P[1][0] - k1 * P[0][0];
    double p11 = P[1][1] - k1 * P[0][1];
    P[0][0] = p00;
    P[0][1] = p01;
    P[1][0] = p10;
    P[1][1] = p11;
}

KalmanFilterBank::KalmanFilterBank(const std::unordered_map<std::string, KalmanConfig>& configs) {
    for (const auto& [name, cfg] : configs) {
        filters_.emplace(name, KalmanFilter(cfg.process_noise, cfg.observation_noise));
    }
}

FusionResult KalmanFilterBank::fuse(const std::unordered_map<std::string, double>& observations) {
    // Initialize trust scores if empty
    for (const auto& [source, price] : observations) {
        if (trust_scores_.find(source) == trust_scores_.end()) {
            trust_scores_[source] = 0.5;  // neutral start
        }
    }

    // Step 1: Predict all filters
    for (auto& [name, filter] : filters_) {
        filter.predict();
    }

    // Step 2: Update each filter with each observation (trust-weighted)
    std::unordered_map<std::string, double> innovations;
    for (const auto& [source, price] : observations) {
        double trust = trust_scores_[source];
        for (auto& [name, filter] : filters_) {
            filter.update(price, trust);
        }

        // Track innovation for trust update
        double avg_prediction = 0.0;
        for (const auto& [name, filter] : filters_) {
            if (filter.initialized()) {
                avg_prediction += filter.price();
            }
        }
        avg_prediction /= filters_.size();
        innovations[source] = price - avg_prediction;
    }

    // Step 3: Fuse filter outputs (inverse-uncertainty weighting)
    double sum_weighted_price = 0.0;
    double sum_weights = 0.0;
    for (const auto& [name, filter] : filters_) {
        if (filter.initialized()) {
            double weight = 1.0 / (filter.uncertainty() + 1e-9);
            sum_weighted_price += filter.price() * weight;
            sum_weights += weight;
        }
    }
    double fused_price = sum_weighted_price / (sum_weights + 1e-9);
    double confidence = 1.0 / (1.0 + std::sqrt(sum_weights / filters_.size()));

    // Step 4: Compute source weights (by trust score)
    std::unordered_map<std::string, double> source_weights;
    double trust_sum = 0.0;
    for (const auto& [source, trust] : trust_scores_) {
        trust_sum += trust;
    }
    for (const auto& [source, trust] : trust_scores_) {
        source_weights[source] = trust / (trust_sum + 1e-9);
    }

    // Step 5: Update trust scores based on innovations
    for (const auto& [source, innovation] : innovations) {
        double avg_uncertainty = 0.0;
        for (const auto& [name, filter] : filters_) {
            if (filter.initialized()) {
                avg_uncertainty += filter.uncertainty();
            }
        }
        avg_uncertainty /= filters_.size();
        update_trust(source, innovation, avg_uncertainty);
    }

    // Step 6: Detect anomalies (any source >3σ from fused)
    bool anomaly_flag = false;
    for (const auto& [source, price] : observations) {
        double z = std::abs(price - fused_price) / (confidence + 1e-9);
        if (z > 3.0) {
            anomaly_flag = true;
            break;
        }
    }

    return {fused_price, confidence, source_weights, anomaly_flag};
}

void KalmanFilterBank::update_trust(const std::string& source, double innovation, double uncertainty) {
    double z_score = std::abs(innovation) / (std::sqrt(uncertainty) + 1e-9);
    double& trust = trust_scores_[source];

    if (z_score < 1.0) {
        trust = std::min(0.95, trust + 0.01);  // increase trust
    } else if (z_score > 3.0) {
        trust = std::max(0.05, trust - 0.05);  // decrease trust
    }
}

} // namespace dominion
